onnxscript/rewriter/ort_fusions/_core.py

```diff
@@ -67,35 +71,42 @@ def fuse_xformers(model: ir.Model) -> tuple[ir.Model, dict[str, int]]:
     fusion_count = dict()
 
     model = _pre_optimize(model)
-    fusion_count["erf_gelu"] = fuse_erfgelu(model)
-    fusion_count["rms_normalization"] = fuse_rms_normalization(model)
-    fusion_count["skip_layer_normalization"] = fuse_skip_layer_normalization(model)
-    fusion_count["skip_rms_normalization"] = fuse_skip_rms_normalization(model)
-    fusion_count["rotary_embedding"] = fuse_rotary_embedding(model)
-    fusion_count["partial_rotary_embedding"] = fuse_partial_rotary_embedding(model)
-    fusion_count["cos_sin_cache"] = fuse_cos_sin_cache(model)
-    fusion_count["sdpa"] = fuse_sdpa(model)
+
+    def fuse(func):
+        return func(model, debug=debug, apply_shape_inference=apply_shape_inference)
+
+    fusion_count["erf_gelu"] = fuse(fuse_erfgelu)
+    fusion_count["rms_normalization"] = fuse(fuse_rms_normalization)
+    fusion_count["skip_layer_normalization"] = fuse(fuse_skip_layer_normalization)
+    fusion_count["skip_rms_normalization"] = fuse(fuse_skip_rms_normalization)
+    fusion_count["rotary_embedding"] = fuse(fuse_rotary_embedding)
+    fusion_count["partial_rotary_embedding"] = fuse(fuse_partial_rotary_embedding)
+    fusion_count["cos_sin_cache"] = fuse(fuse_cos_sin_cache)
+    fusion_count["sdpa"] = fuse(fuse_sdpa)
     # Optimize to avoid trying multiple attention-based fusions
-    fusion_count["mha"] = fuse_mha(model)
+    fusion_count["mha"] = fuse(fuse_mha)
     if fusion_count["mha"] == 0:
         # If no MHA fusion was applied, we can try the GQA fusion.
         # and avoid trying the attention fusion.
-        fusion_count["gqa"] = fuse_gqa(model)
-        fusion_count["packed_qkv_for_gqa"] = fuse_qkv_gqa(model)
+        fusion_count["gqa"] = fuse(fuse_gqa)
+        fusion_count["packed_qkv_for_gqa"] = fuse(fuse_qkv_gqa)
         fusion_count["attention"] = 0
     else:
-        fusion_count["attention"] = fuse_attention(model)
+        fusion_count["attention"] = fuse(fuse_attention)
         fusion_count["gqa"] = 0
-    fusion_count["gelu"] = fuse_gelu(model)
-    fusion_count["bias_gelu"] = fuse_bias_gelu(model)
+    fusion_count["gelu"] = fuse(fuse_gelu)
+    fusion_count["bias_gelu"] = fuse(fuse_bias_gelu)
     # Finally: inline any intermediate fusion functions introduced that were not
     # consumed by other fusions, and eliminate any remaining unused nodes.
     optimize(model)
     return model, fusion_count
 
 
 def optimize_for_ort(
-    model: ir.Model, config_name: str | None = None
+    model: ir.Model,
+    config_name: str | None = None,
+    debug: bool = False,
+    apply_shape_inference: bool = False,
```

Comment:
Is it necessary to expose this to an external user? Or, can the implementation  do the right thing?

Reply:
I mean the "apply_shape_inference" flag.

Reply:
I think keeping the flag internal would be better, although we might need to determine which fusions would be need/benefit from having the shape inference call after fusion. For example, SDPA -> shape_inf -> MHA would be a good candidate for this, given SDPA could create some nodes in the rewrite for the only_q_pre_scale pattern